### PR TITLE
NEW: Add NuClick architecture

### PR DESCRIPTION
**Pull request by mostafajahanifar**
This PR adds NuClick architecture, as described in the paper below, to the toolbox.

> Koohbanani, Navid Alemi, et al. "NuClick: a deep learning framework for interactive segmentation of microscopic images." Medical Image Analysis 65 (2020): 101771.

### CODE REVIEW

File: tiatoolbox/models/architecture/nuclick.py [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+        kernel_size (int): Size of the kernel in the convolution layer.
+        strds (int): Size of the stride in the convolution layer.
+        use_bias (bool): Whether to use bias in the convolution layer.
+        dilatation_rate (int): Dilatation rate in the convolution layer.
+        activation (str): Name of the activation function to use.
+        do_batchnorm (bool): Whether to do batch normalization after the
+        convolution layer.
+    Returns:
+        model (torch.nn.Module): a pytorch model.
+    """
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:40]:
```suggestion

    """
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+
+    def __init__(
+        self,
+        num_input_channels,
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:44]: Add typehints
```suggestion
        num_input_channels: int,
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+        )
+
+    def forward(self, input_tensor):
+        """Logic for using layers defined in init.
+        This method defines how layers are used in forward operation.
+        Args:
+            input_tensor (torch.Tensor): Input, the tensor is of the shape NCHW.
+        Returns:
+            output (torch.Tensor): The inference output.
+        """
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:78]:
```suggestion

        """
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+            num_output_channels,
+            kernel_size,
+            strds,
+            use_bias,
+            dilatation_rate,
+            activation,
+            do_batchnorm,
+        )
+
+    def forward(self, input_tensor):
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:71]:
```suggestion
    def forward(self, input_tensor: torch.Tensor)-> torch.Tensor:
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+    Args:
+        num_input_channels (int): Number of channels in input.
+        num_output_channels (int): Number of channels in output.
+        kernel_sizes (list): Size of the kernel in each convolution layer.
+        strds (int): Size of stride in the convolution layer.
+        use_bias (bool): Whether to use bias in the convolution layer.
+        dilatation_rates (list): Dilation rate for each convolution layer.
+        activation (str): Name of the activation function to use.
+    Returns:
+        model (torch.nn.Module): a pytorch model.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:130]:
```suggestion
        torch.nn.Module: 
            A PyTorch model.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+            self.activation = nn.Sigmoid()
+
+    def forward(self, input_tensor):
+        """Logic for using layers defined in init.
+        This method defines how layers are used in forward operation.
+        Args:
+            input_tensor (torch.Tensor): Input, the tensor is of the shape NCHW.
+        Returns:
+            output (torch.Tensor): The inference output.
+        """
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:266]:
```suggestion

        """
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+    NuClick takes an RGB image patch along with an inclusion and an exclusion map.
+    Args:
+        num_input_channels (int): Number of channels in input.
+        num_output_channels (int): Number of channels in output.
+    Returns:
+        model (torch.nn.Module): a pytorch model.
+    Examples:
+        >>> # instantiate a NuClick model for interactive nucleus segmentation.
+        >>> NuClick(num_input_channels = 5, num_output_channels = 1)
+    """
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:287]:
```suggestion

    """
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+        )
+
+    def forward(self, imgs: torch.Tensor):
+        """Logic for using layers defined in init.
+        This method defines how layers are used in forward operation.
+        Args:
+            imgs (torch.Tensor): Input images, the tensor is of the shape NCHW.
+        Returns:
+            output (torch.Tensor): The inference output.
+        """
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:451]:
```suggestion

        """
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+
+        up9 = torch.cat([self.conv_transpose_5(conv8), conv1], dim=1)
+        conv9 = self.conv_block_2(up9)
+
+        return self.conv_block_3(conv9)
+
+    @staticmethod
+    def generate_inst_dict(pred_mask, bounding_boxes):
+        """To collect instance information and store it within a dictionary.
+        Args:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:495]:
```suggestion

        Args:
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 10]
```diff
@@ -0,0 +1,577 @@
+"""Define orginal NuClick architecture#
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:1]:
```suggestion
"""Define original NuClick architecture
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+
+Koohbanani, N. A., Jahanifar, M., Tajadin, N. Z., & Rajpoot, N. (2020).
+NuClick: a deep learning framework for interactive segmentation of microscopic images.
+Medical Image Analysis, 65, 101771.
+"""
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:6]:
```suggestion

"""
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+            dilatation_rate,
+            activation,
+            do_batchnorm,
+        )
+
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+        """Logic for using layers defined in init.
+        This method defines how layers are used in forward operation.
+        Args:
+            input_tensor (torch.Tensor): Input, the tensor is of the shape NCHW.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:77]:
```suggestion
            input_tensor (torch.Tensor): 
                    Input, the tensor is of the shape NCHW.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 13]
```diff
@@ -0,0 +1,577 @@
...
+        Returns:
+            output (torch.Tensor): The inference output.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:79]:
```suggestion
            output (torch.Tensor): 
                        The inference output.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 14]
```diff
@@ -0,0 +1,577 @@
...
+)
+
+from tiatoolbox.models.models_abc import ModelABC
+from tiatoolbox.utils import misc
+
+bn_axis = 1
+
+
+class ConvBnRelu(nn.Module):
+    """Convolution -> Batch Normalization -> ReLu/Sigmoid
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:27]:
```suggestion
        """Performs Convolution, Batch Normalization and activation.
    
    Convolution -> Batch Normalization -> ReLu/Sigmoid
    
    Args:
        num_input_channels (int): 
            Number of channels in input.
        num_output_channels (int): 
            Number of channels in output.
        kernel_size (int): 
            Size of the kernel in the convolution layer.
        strds (int): 
            Size of the stride in the convolution layer.
        use_bias (bool): 
            Whether to use bias in the convolution layer.
        dilatation_rate (int): 
            Dilatation rate in the convolution layer.
        activation (str): 
            Name of the activation function to use.
        do_batchnorm (bool): 
            Whether to do batch normalization after the convolution layer.
            
    Returns:
        model (torch.nn.Module): a pytorch model.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 15]
```diff
@@ -0,0 +1,577 @@
...
+            kernel_size,
+            strds,
+            use_bias,
+            dilatation_rate,
+            activation,
+            do_batchnorm,
+        )
+
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+        """Logic for using layers defined in init.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:74]:
```suggestion
        """Logic for using layers defined in init.
        
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+        This method defines how layers are used in forward operation.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:75]:
```suggestion
        This method defines how layers are used in forward operation.
        
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 17]
```diff
@@ -0,0 +1,577 @@
...
+        in_channels,
+        out_channels,
+        kernel_size,
+        strds,
+        use_bias,
+        dilatation_rate,
+        activation,
+        do_batchnorm,
+    ):
+        """Function for acquire the convolutional block.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:95]:
```suggestion
        """Function for acquire the convolutional block.
        
```

Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:95]:
```suggestion
        """Function for acquire the convolutional block.
        
        Args:
            in_channels (int): 
                Number of channels in input.
            out_channels (int): 
                Number of channels in output.
            kernel_size (list): 
                Size of the kernel in the acquired convolution block.
            strds (int): 
                Size of stride in the convolution layer.
            use_bias (bool): 
                Whether to use bias in the convolution layer.
            dilatation_rates (list): 
                Dilation rate for each convolution layer.
            activation (str): 
                Name of the activation function to use.
            do_batchnorm (bool): 
                Whether to do batch normalization after the convolution layer.
                
        Returns:
            torch.nn.Sequential: a pytorch layer
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 18]
```diff
@@ -0,0 +1,577 @@
...
+        return self.conv_bn_relu(input_tensor)
+
+    @staticmethod
+    def get_block(
+        in_channels,
+        out_channels,
+        kernel_size,
+        strds,
+        use_bias,
+        dilatation_rate,
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:91]:
```suggestion
        dilation_rate, ?
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 19]
```diff
@@ -0,0 +1,577 @@
...
+        activation,
+        do_batchnorm,
+    ):
+        """Function for acquire the convolutional block.
+        Args:
+            in_channels (int): Number of channels in input.
+            out_channels (int): Number of channels in output.
+            kernel_size (list): Size of the kernel in the acquired convolution block.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:99]:
```suggestion
            kernel_size (Union[int, Tuple[int, int]]): Size of the kernel in the acquired convolution block.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 20]
```diff
@@ -0,0 +1,577 @@
...
+            layers.append(nn.BatchNorm2d(num_features=out_channels, eps=1.001e-5))
+
+        if activation == "relu":
+            layers.append(nn.ReLU())
+
+        return nn.Sequential(*layers)
+
+
+class MultiscaleConvBlock(nn.Module):
+    """Multiscale convolution block
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:135]:
```suggestion
    """Defines Multiscale convolution block.
    
    Defines four convolution layers.
    
    Args:
        num_input_channels (int): 
            Number of channels in input.
        num_output_channels (int): 
            Number of channels in output.
        kernel_sizes (list): 
            Size of the kernel in each convolution layer.
        strds (int): 
            Size of stride in the convolution layer.
        use_bias (bool): 
            Whether to use bias in the convolution layer.
        dilatation_rates (list): 
            Dilation rate for each convolution layer.
        activation (str): 
            Name of the activation function to use.
            
    Returns:
        torch.nn.Module:
            A PyTorch model.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 21]
```diff
@@ -0,0 +1,577 @@
...
+    Defines four convolution layers.
+    Args:
+        num_input_channels (int): Number of channels in input.
+        num_output_channels (int): Number of channels in output.
+        kernel_sizes (list): Size of the kernel in each convolution layer.
+        strds (int): Size of stride in the convolution layer.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:141]: why not be consistent with stride?

**Comment by mostafajahanifar** [comment]
> @kesixu @mostafajahanifar
> 
> The nuclick architecture file needs to have docstring consistent with the rest of the repo.
> 
> It needs to be fixed. The remaining three files are good.

Dear Shan,

Thanks for your comments and commit suggestions. I've incorporated all of them in the PR. Please check if you like the changes.

File: tiatoolbox/models/architecture/nuclick.py [hunk 22]
```diff
@@ -0,0 +1,636 @@
...
+            Number of channels in output.
+        kernel_size (int):
+            Size of the kernel in all convolution layers.
+        strides (int):
+            Size of the stride in all convolution layers.
+        use_bias (bool):
+            Whether to use bias in the convolution layers.
+        dilation_rate (int):
+            Dilation rate in all convolution layers.
+    Returns:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:274]:
```suggestion

    Returns:
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 23]
```diff
@@ -0,0 +1,636 @@
...
+            do_batchnorm=True,
+        )
+
+        self.activation = nn.ReLU()
+
+    def forward(self, input_tensor):
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+        Args:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:318]:
```suggestion

        Args:
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 24]
```diff
@@ -0,0 +1,636 @@
...
+            input_tensor (torch.Tensor):
+                Input, the tensor is of the shape NCHW.
+        Returns:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:321]:
```suggestion

        Returns:
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 25]
```diff
@@ -0,0 +1,636 @@
...
+
+        This contains logic for forward operation as well as batch i/o
+        aggregation.
+
+        Args:
+            model (nn.Module): PyTorch defined model.
+            batch_data (ndarray): a batch of data generated by
+                torch.utils.data.DataLoader.
+            on_gpu (bool): Whether to run inference on a GPU.
+        Returns:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:621]:
```suggestion

        Returns:
```

File: tests/models/test_arch_nuclick.py [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    """Tests for nuclick."""
+    # convert to pathlib Path to prevent wsireader complaint
+    tile_path = pathlib.Path(remote_sample("patch-extraction-vf"))
+    img = imread(tile_path)
+
+    _pretrained_path = f"{tmp_path}/weights.pth"
+    fetch_pretrained_weights("nuclick_original-pannuke", _pretrained_path)
+
+    # test creation
+    model = NuClick(num_input_channels=5, num_output_channels=1)
```
Comment (shaneahmed) [inline tests/models/test_arch_nuclick.py:28]:
```suggestion
    model = NuClick(num_input_channels=5, num_class=1)
```

File: tests/models/test_arch_nuclick.py [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+
+    patch = np.float32(patch) / 255.0
+    patch = np.moveaxis(patch, -1, 0)
+    batch = np.concatenate(
+        (patch, inclusion_map[np.newaxis, ...], exclusion_map[np.newaxis, ...]), axis=0
+    )
+
+    batch = torch.from_numpy(batch[np.newaxis, ...])
+
+    model = NuClick(num_input_channels=5, num_output_channels=1)
```
Comment (shaneahmed) [inline tests/models/test_arch_nuclick.py:49]:
```suggestion
    model = NuClick(num_input_channels=5, num_class=1)
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 28]
```diff
@@ -0,0 +1,640 @@
+"""Define original NuClick architecture
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:1]:
```suggestion
"""Defines original NuClick architecture
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 29]
```diff
@@ -0,0 +1,640 @@
...
+
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                    Input, the tensor is of the shape NCHW.
+        Returns:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:90]:
```suggestion
        
        Returns:
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 30]
```diff
@@ -0,0 +1,640 @@
...
+        return nn.Sequential(*layers)
+
+
+class MultiscaleConvBlock(nn.Module):
+    """Defines Multiscale convolution block.
+
+    Args:
+        num_input_channels (int):
+            Number of channels in input.
+        num_output_channels (int):
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:163]:
```suggestion
        num_class (int):
```

Shouldn't we call it num_class as this is the number of output classes?
If you think we should keep it num_output_channels then please refactor this in other models as well.

File: tiatoolbox/models/architecture/nuclick.py [hunk 31]
```diff
@@ -0,0 +1,640 @@
...
+        self,
+        num_input_channels: int,
+        kernel_sizes: Union[Tuple[int, int], np.ndarray],
+        dilation_rates: Union[Tuple[int, int], np.ndarray],
+        num_output_channels: int = 32,
+        strides: Union[Tuple[int, int], np.ndarray] = (1, 1),
+        activation: str = "relu",
+        use_bias: bool = False,
+    ):
+
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:192]:
```suggestion
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 32]
```diff
@@ -0,0 +1,640 @@
...
+            num_output_channels=num_output_channels,
+            kernel_size=kernel_sizes[3],
+            strides=strides,
+            activation=activation,
+            use_bias=use_bias,
+            dilation_rate=(dilation_rates[3], dilation_rates[3]),
+        )
+
+    def forward(self, input_map):
+        """Logic for using layers defined in init.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:236]:
```suggestion
        """Logic for using layers defined in MultiscaleConvBlock init.
```

Please correct this if I am wrong?

File: tiatoolbox/models/architecture/nuclick.py [hunk 33]
```diff
@@ -0,0 +1,640 @@
...
+        conv1 = self.conv_block_1(conv0)
+        conv2 = self.conv_block_2(conv0)
+        conv3 = self.conv_block_3(conv0)
+        conv4 = self.conv_block_4(conv0)
+
+        return torch.cat([conv1, conv2, conv3, conv4], dim=bn_axis)
+
+
+class ResidualConv(nn.Module):
+    """Residual Convolution block
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:259]:
```suggestion
    """Residual Convolution block.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 34]
```diff
@@ -0,0 +1,640 @@
...
+        strides (int):
+            Size of the stride in all convolution layers.
+        use_bias (bool):
+            Whether to use bias in the convolution layers.
+        dilation_rate (int):
+            Dilation rate in all convolution layers.
+
+    Returns:
+        model (torch.nn.Module):
+            a pytorch model.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:277]:
```suggestion
            A pytorch model.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 35]
```diff
@@ -0,0 +1,640 @@
...
+
+    """
+
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:280]:
```suggestion
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 36]
```diff
@@ -0,0 +1,640 @@
...
+            activation="None",
+            use_bias=use_bias,
+            dilation_rate=dilation_rate,
+            do_batchnorm=True,
+        )
+
+        self.activation = nn.ReLU()
+
+    def forward(self, input_tensor):
+        """Logic for using layers defined in init.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:316]:
```suggestion
        """Logic for using layers defined in ResidualConv init.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 37]
```diff
@@ -0,0 +1,640 @@
...
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                Input, the tensor is of the shape NCHW.
+
+        Returns:
+            output (torch.Tensor): The inference output.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:325]:
```suggestion
            output (torch.Tensor): 
                The inference output.
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 38]
```diff
@@ -0,0 +1,640 @@
...
+class NuClick(ModelABC):
+    """NuClick Architecture.
+
+    NuClick is used for interactive nuclei segmentation.
+    NuClick takes an RGB image patch along with an inclusion and an exclusion map.
+
+    Args:
+        num_input_channels (int):
+            Number of channels in input.
+        num_output_channels (int):
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:344]:
```suggestion
        num_class (int):
```

File: tiatoolbox/models/architecture/nuclick.py [hunk 39]
```diff
@@ -0,0 +1,640 @@
...
+
+        self.conv_transpose_5 = nn.ConvTranspose2d(
+            in_channels=64,
+            out_channels=32,
+            kernel_size=2,
+            stride=(2, 2),
+        )
+
+    def forward(self, imgs: torch.Tensor):
+        """Logic for using layers defined in init.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/nuclick.py:513]:
```suggestion
        """Logic for using layers defined in NuClick init.
```

**Review by shaneahmed — CHANGES_REQUESTED** [review]
Thanks @mostafajahanifar  and @kesixu. Please can you address some minor comments. We should be ready to merge it then.

File: tests/models/test_arch_nuclick.py (hunk 26 above)
Comment (mostafajahanifar) [reply tests/models/test_arch_nuclick.py:28]: Thanks for the suggestion but I think it's better to keep it as it is to be in line with other networks (UNet, HoVerNet) implementation in the toolbox.

File: tests/models/test_arch_nuclick.py (hunk 27 above)
Comment (mostafajahanifar) [reply tests/models/test_arch_nuclick.py:49]: Thanks for the suggestion but I think it's better to keep it as it is to be in line with other networks (UNet, HoVerNet) implementation in the toolbox.

File: tiatoolbox/models/architecture/nuclick.py (hunk 30 above)
Comment (mostafajahanifar) [reply tiatoolbox/models/architecture/nuclick.py:163]: Thanks for the suggestion but I think it's better to keep it as it is to be in line with other networks (UNet) implementation in the toolbox.

File: tests/models/test_arch_nuclick.py (hunk 26 above)
Comment (John-P) [reply tests/models/test_arch_nuclick.py:28]: Hmm yes I think it changing it, then it should be everywhere. Consistency > brevity.

File: tiatoolbox/models/architecture/nuclick.py (hunk 30 above)
Comment (mostafajahanifar) [reply tiatoolbox/models/architecture/nuclick.py:164]: The number of output classes may be inferred differently with different networks. Some might use a FC layer with the same number of nodes as "number of output classes" to classify/detect/segment the objects in the image, other methods might use a number of output channels to identify a single class (instance segmentation where we have marker and mask), and some methods that use only one channel to detect/segment different "number of output classes".

My point is that these are valid and invalid. I agree that it should be consistent through the toolbox but might be impossible due to the nature of different models. Anyway, I tried to keep this network architecture in line with UNet implementation as much as possible.

**Comment by mostafajahanifar** [comment]
Dear @shaneahmed 

I revised the code to make the input to all models consistent with `num_output_channels` argument. In particular, I changed MicroNet argument to `num_output_channels` but for MapDe changed it to `num_classes` because argument is actually referring to number of cell classes based on which number of output channels is generated ( `num_output_channels=2*num_classes` ).

I have also refactor all the models to use predefined (base class) `preproc` method instead of `_tranform` method. And made some other changes to avoid `[PYL-W0221]` pylint error.

**Comment by shaneahmed** [comment]
> Dear @shaneahmed
> 
> I revised the code to make the input to all models consistent with `num_output_channels` argument. In particular, I changed MicroNet argument to `num_output_channels` but for MapDe changed it to `num_classes` because argument is actually referring to number of cell classes based on which number of output channels is generated ( `num_output_channels=2*num_classes` ).
> 
> I have also refactor all the models to use predefined (base class) `preproc` method instead of `_tranform` method. And made some other changes to avoid `[PYL-W0221]` pylint error.

Thanks @mostafajahanifar Please can you check the errors?

**Review by shaneahmed — APPROVED** [review]

**Pull request merged**